BCnEnc.Net/Encoder/Bc4BlockEncoder.cs

```diff
@@ -1,146 +1,192 @@
-				if (luminanceAsRed) {
+			for (var i = 0; i < 16; i++)
+			{
+				if (luminanceAsComponent)
+				{
 					colors[i] = (byte)(new ColorYCbCr(pixels[i]).y * 255);
 				}
-				else {
-					colors[i] = pixels[i].R;
+				else
+				{
+					switch (component)
+					{
+						case Bc4Component.R:
+							colors[i] = pixels[i].R;
+							break;
+
+						case Bc4Component.G:
+							colors[i] = pixels[i].G;
+							break;
+
+						case Bc4Component.B:
+							colors[i] = pixels[i].B;
+							break;
+
+						case Bc4Component.A:
+							colors[i] = pixels[i].A;
+							break;
+					}
 				}
 			}
- 			switch (quality) {
+			switch (quality)
+			{
 				case CompressionQuality.Fast:
-					return FindRedValues(output, colors, 3);
+					return FindComponentValues(output, colors, 3);
 				case CompressionQuality.Balanced:
-					return FindRedValues(output, colors, 4);
+					return FindComponentValues(output, colors, 4);
 				case CompressionQuality.BestQuality:
-					return FindRedValues(output, colors, 8);
+					return FindComponentValues(output, colors, 8);
 
 				default:
 					throw new ArgumentOutOfRangeException(nameof(quality), quality, null);
 			}
 		}
 
-		public override GlInternalFormat GetInternalFormat() {
+		public override GlInternalFormat GetInternalFormat()
+		{
 			return GlInternalFormat.GlCompressedRedRgtc1Ext;
 		}
 
-		public override GlFormat GetBaseInternalFormat() {
+		public override GlFormat GetBaseInternalFormat()
+		{
 			return GlFormat.GlRed;
 		}
 
-		public override DxgiFormat GetDxgiFormat() {
+		public override DxgiFormat GetDxgiFormat()
+		{
 			return DxgiFormat.DxgiFormatBc4Unorm;
 		}
 
 		#region Encoding private stuff
 
-		private static Bc4Block FindRedValues(Bc4Block colorBlock, byte[] pixels, int variations) {
+		private static Bc4Block FindComponentValues(Bc4Block colorBlock, byte[] pixels, int variations)
+		{
 
 			//Find min and max alpha
 			byte min = 255;
 			byte max = 0;
 			var hasExtremeValues = false;
-			for (var i = 0; i < pixels.Length; i++) {
-				if (pixels[i] < 255 && pixels[i] > 0) {
+			for (var i = 0; i < pixels.Length; i++)
+			{
+				if (pixels[i] < 255 && pixels[i] > 0)
+				{
 					if (pixels[i] < min) min = pixels[i];
 					if (pixels[i] > max) max = pixels[i];
 				}
-				else {
+				else
+				{
 					hasExtremeValues = true;
 				}
 			}
 
 
-			int SelectIndices(ref Bc4Block block) {
+			int SelectIndices(ref Bc4Block block)
+			{
 				var cumulativeError = 0;
-				var c0 = block.Red0;
-				var c1 = block.Red1;
+				var c0 = block.Endpoint0;
+				var c1 = block.Endpoint1;
 				var colors = c0 > c1
 					? stackalloc byte[] {
-						c0,
-						c1,
-						(byte) (6 / 7.0 * c0 + 1 / 7.0 * c1),
-						(byte) (5 / 7.0 * c0 + 2 / 7.0 * c1),
-						(byte) (4 / 7.0 * c0 + 3 / 7.0 * c1),
-						(byte) (3 / 7.0 * c0 + 4 / 7.0 * c1),
-						(byte) (2 / 7.0 * c0 + 5 / 7.0 * c1),
-						(byte) (1 / 7.0 * c0 + 6 / 7.0 * c1),
-					}
+					c0,
+					c1,
+					(byte)(6 / 7.0 * c0 + 1 / 7.0 * c1),
```

Comment:
Could use the new Interpolation functions here too.

Reply:
Yh, all encoders seem to do interpolations on their own and I had the same though of using the Interpolation class in such events. I left it out for now, since this commit was meant to add ATC first and foremost.